manuscript.Rmd

```diff
@@ -302,8 +308,95 @@ The cloud is the natural environment of containers, and becomes the go-to mechan
+
+At T-Mobile, R is used to power real-time services that are called over a million times a day. T-Mobile created a set of neural network machine learning natural language processing models to help customer care agents manage text-based messages for customers. The models are convolutional neural networks that use the RStudio Keras package and are built on top of a Rocker Docker image. Since the models power tools for agents and customers, they need to have extremely high uptime and reliability--and they found that R is able to perform. The AI @ T-Mobile team open sourced [their docker images](https://github.com/tmobile/r-tensorflow-api  which are available on GitHub.
+
+## Deployment to the cloud [@hongooi]
+
+While Plumber provides the low-level infrastructure for turning a statistical model into a predictive service, for production purposes it is usually necessary to take into account considerations such as scalability, reliability and ease of management. A single container is limited in the volume of requests it can service. Furthermore, if the container goes down, or is occupied with processing a computationally intensive task, the service becomes unavailable until it is restarted.
+
+These issues can be mitigated by deploying not a single container, but a _cluster_ of containers, orchestrated as a single deployment. The deployment system of choice at the time of writing is [Kubernetes](https://kubernetes.io). This allows you to deploy a single image to a pool of servers, which can then serve requests in parallel. Client-side requests are made to the cluster ingress endpoint, which then redirects them to individual servers. The orchestrator can identify cluster nodes that have failed, and restart them automatically; it can also detect when there is no capacity to service more requests, and start additional servers and containers as needed.
+
+While these load-balancing and autoscaling properties of a Kubernetes cluster are invaluable, they are still limited by the hardware resources that it has access to. A cluster cannot scale out if it has run out of servers. This issue in turn can be addressed by deploying not to on-premise servers but to the cloud, where effectively unlimited computing resources are available (as long as one can pay for them). From a management perspective, cloud deployments have the additional advantage that hardware maintenance becomes the responsibility of the cloud provider, eliminating another pain point. Orchestrator services are available from all the major cloud service providers, including Amazon Web Services (AWS), Google Cloud Platform, and Microsoft Azure.
```

Comment:
Same here: This is all helpful, but potentially a bit too extensive for the scope of the article. I'll work on a revision and let you know.